presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java

```diff
@@ -73,6 +89,334 @@ public void setUp()
+        @ScalarFunctionConstantStats(distinctValuesCount = 2.0, nullFraction = 0.0)
+        public static boolean customIsNullVarchar(@SqlNullable @SqlType("varchar(x)") Slice slice)
+        {
+            return slice == null;
+        }
+
+        @ScalarFunction(value = "custom_is_null", calledOnNullInput = true)
+        @SqlType(StandardTypes.BOOLEAN)
+        @ScalarFunctionConstantStats(distinctValuesCount = 2.0, nullFraction = 0.0)
+        public static boolean customIsNullBigint(@SqlNullable @SqlType(StandardTypes.BIGINT) Long value)
+        {
+            return value == null;
+        }
+
+        @ScalarFunction(value = "custom_str_len")
+        @SqlType(StandardTypes.BIGINT)
+        @LiteralParameters("x")
+        @ScalarFunctionConstantStats(minValue = 0)
+        public static long customStrLength(
+                @ScalarPropagateSourceStats(
+                        propagateAllStats = false,
+                        nullFraction = StatsPropagationBehavior.USE_SOURCE_STATS,
+                        distinctValuesCount = StatsPropagationBehavior.USE_TYPE_WIDTH_VARCHAR,
+                        maxValue = StatsPropagationBehavior.USE_TYPE_WIDTH_VARCHAR) @SqlType("varchar(x)") Slice value)
+        {
+            return value.length();
+        }
+
+        @ScalarFunction(value = "custom_str_edit_distance")
+        @SqlType(StandardTypes.BIGINT)
+        @LiteralParameters({"x", "y"})
+        @ScalarFunctionConstantStats(minValue = 0)
+        public static long customStrEditDistance(
+                @ScalarPropagateSourceStats(
+                        propagateAllStats = false,
+                        nullFraction = StatsPropagationBehavior.USE_MAX_ARGUMENT,
+                        distinctValuesCount = StatsPropagationBehavior.MAX_TYPE_WIDTH_VARCHAR,
+                        maxValue = StatsPropagationBehavior.MAX_TYPE_WIDTH_VARCHAR) @SqlType("varchar(x)") Slice str1,
+                @SqlType("varchar(y)") Slice str2)
+        {
+            return 100;
+        }
+
+        @ScalarFunction(value = "custom_prng", calledOnNullInput = true)
+        @SqlType(StandardTypes.BIGINT)
+        @LiteralParameters("x")
+        @ScalarFunctionConstantStats(nullFraction = 0)
+        public static long customPrng(
+                @SqlNullable
+                @ScalarPropagateSourceStats(
+                        propagateAllStats = false,
+                        distinctValuesCount = StatsPropagationBehavior.ROW_COUNT,
+                        minValue = StatsPropagationBehavior.USE_SOURCE_STATS) @SqlType(StandardTypes.BIGINT) Long min,
+                @ScalarPropagateSourceStats(
+                        propagateAllStats = false,
+                        maxValue = StatsPropagationBehavior.USE_SOURCE_STATS
+                ) @SqlNullable @SqlType(StandardTypes.BIGINT) Long max)
+        {
+            return (long) ((Math.random() * (max - min)) + min);
+        }
+    }
+
+    @Test
```

Comment:
Can you add negative tests -
1. NDV value set in annotation to higher than output row count
2. Average row size is set to greater than the actual width of the type
3. max is greater than min
4. null fraction is negative or more than `1`

For some of the cases (negative values, out of bound values) would the stats propagation fail to register and would we get the existing do-not-propagate  stats behavior ?

Reply:
Yes, only exception is average row size which is upper bounded to return type width of the function (if known).

Reply:
So, I am thinking of changing this slightly, 
1. If average row size is provided via `ScalarFunctionConstantStats` annotation, then we allow even if the size is greater than functions return type width.
2. If average row size is provided via one of the `StatsPropagationBehavior` values, then we upper bound it to functions return type width - if available.
3. If both 1 and 2 is unknown, then we default it to functions return type width if available.

This way the function developer has greater control.

Reply:
For distinct values count, upper bounding it to row count is as good as unknown. Therefore, the approach here is, when distinctValuesCount is greater than row count and is provided via annotation we set it to unknown. 
A function developer has full control here, for example developer can choose to upper bound or not by selecting the appropriate `StatsPropagationBehavior` value.